docs/devel/infrastructure-interaction.md

```diff
@@ -0,0 +1,85 @@
+
+Between minions, GCE PDs, and GCE external load balancers, all of these have different implementations with little coherence for non-gce configurations. So thinking generally, how should kube and infrastructure functionality work.  If I’m looking at the kube cluster from that 3rd infrastructure interaction perspective (and possibly even the 1st perspective) I think we must view that underlying functionality as pets, not cattle.  That’s just a perspective I’m bringing into this discussion.
+
+To decouple kube and the infrastructure I believe that our model of interaction should require three distinct steps and three distinct parties interacting.
+
+**Design Idea**
+
+There are three parties who need to operate on the system in my mind.  They may be actual people running things on a command line or automated tools doing something similar.  That’s not really relevant.
+
+1. Infrastructure admin.  The person with the permission and ability to make changes to the infrastructure.
+2. Kube admin.  The person with the permission to change how the kube cluster itself operates.
+3. Kube user.  The person who just wants to write and run an app.
+
+My actual design decoupling can be broken down into 3 steps.
+
+1. The “infrastructure admin”, needs to make these external functions available to kube, potentially via a cloud provider interface similar to how minions are handled today.  We would want something similar for storage and external reachability functionality.  It seems likely that we may want to decompose the cloud_provider into multiple smaller providers.  ‘minion_provider’, ‘storage_provider’, etc.  But that is a specific detail to work out per infrastructure concept.
+
+2. The “kube admin” needs to explicitly tell the cluster to use the functionality exposed by the infrastructure. In the case of minions this defaults to “*” but we have the –minion_regexp. That means minions sorta fit this model. GCE PDs fit as you have to specify the GCEPersistentDisk.  The specifics of how the kube admin tells the cluster to use the functionality will likely be functionality specific.  (Although I’ll give my thoughts on some of these things below)
+
+3. “Something” needs to watch the cluster to see how the resource is supposed to be used and make that happen. Minions fall down here, it's more that the apiserver reaches out to them, but I think that’s an ok exception.  The model still mostly fits for GCE PDs. The kubelet watches the pod description and from the volume info in the pod definition it mounts and actually uses the storage. I think a different, but very similar, pattern should be generalized and should work well for externally reachable functionality (load balancers, external IP, etc) and other storage mechanisms.
```

Comment:
Not sure what you mean about minions/nodes. We're working on a node controller to oversee them, and also working on self-registration, bring-your-own-node, and node auto-configuration for cluster (manual and auto-) scaling scenarios. In the future, I'd like to see something call out to request nodes on demand. What we currently have for PD doesn't actually allocate the PD on demand, which I'd also like to see happen.